Add Compliance Build to `main`

scbedd opened the pull request:
The code present in this PR powers two builds.

- [assets - compliance - batch](https://dev.azure.com/azure-sdk/internal/_build?definitionId=6850)
  - Scheduled.
  - Picks up everything from last run to now
- [assets- compliance](https://dev.azure.com/azure-sdk/internal/_build?definitionId=6802&_a=summary)
  - Single tag validation.
  - As tags are created

The only thing left (aside from feedback) is to swap the build to pull the tags from the github graph, versus parsing `git log`.

scbedd committed:
add compliance job

scbedd committed:
run the compliance job against tags

scbedd committed:
update parameter title

scbedd committed:
str -> string

scbedd committed:
checkout tag

scbedd committed:
clone into sources

scbedd committed:
small difference again

scbedd committed:
script needs to be multiline

scbedd committed:
we need to actually get the credscan result analysis

scbedd committed:
compliance job now pulls suppressions from main

scbedd committed:
re-add the target commit / tag paramter, with fallback to self

scbedd committed:
Merge branch 'main' into feature/add-compliance-build

scbedd committed:
adding the necessary bits and bobs to enable matrix generation so we can multiplex credscan across the whole run

scbedd committed:
prepare this whole thing to run

scbedd committed:
can we disable a parameter being required?

scbedd committed:
remove required

scbedd committed:
handle defaults a bit better

scbedd committed:
handle defaults even better

scbedd committed:
surround parameter in quotes

scbedd committed:
adding sparse checkout so that we still have the tag script

scbedd committed:
try this

scbedd committed:
rename some stuff. why is it still breaking on checkout?

scbedd committed:
explicitly reference the location

scbedd committed:
fix the checkout

scbedd committed:
why does this basic thing not work?

scbedd committed:
trying to call into the sub function now

scbedd committed:
proper job syntax now

scbedd committed:
add a parameter just to satisfy the image

scbedd committed:
add required parameters

scbedd committed:
add psmodule helpers

scbedd committed:
ensure the dependency relationship is correct

scbedd committed:
ok this should be fixed now. I was generating a 0 indexed option, which crashed create job matric

scbedd committed:
batching is working properly! time to get this going for everything

scbedd committed:
missing | on the pwsh: task shortcut

scbedd committed:
save progress

scbedd committed:
default to last 2 hours.

scbedd committed:
suppression file reference

scbedd committed:
ensure we don't run the matrix job if the tag variable is unpopulated

scbedd committed:
slight refactor to make credscan_batch a stage, and to allow it to skip properly when the tag is not defined

scbedd committed:
save this thing

scbedd committed:
ensure that our paths are straightforward. upload credscan artifacts as artifact. add a lineitem to dump all the content in case of failed

scbedd committed:
ensure that the name of the credscan artifact is totally g2g

scbedd committed:
allow this to run against a specific commit as well as self

scbedd committed:
allow self to explicitly work here

scbedd committed:
trigger condition properly updated

scbedd committed:
dump what we're seeing to understand how this fails

scbedd committed:
ensure the single run works as expected

scbedd committed:
ensure matrix doesn't run for no results

scbedd committed:
proper escaping!

scbedd committed:
how about handle null tag

scbedd committed:
handle when no tags come back

scbedd committed:
extend the durations to 6 hours instead of 2

scbedd committed:
why are we not seeing what we expect?

scbedd committed:
credscan shouldn't fail the way it is

scbedd committed:
remove steps: which was accidentally added

scbedd committed:
publish artifact referenced with full name

scbedd committed:
only publish the credscan artifact if it fails

scbedd committed:
credscan shouldn't trip up on the credscan suppression file

scbedd committed:
ensure we always get the proper folder for scanning

scbedd committed:
the credscan suppression file is malformed

scbedd committed:
maybe finally figure out the credscan suppression failure

scbedd committed:
ensure that .env files are always ignored.

scbedd committed:
ensure that this thing will run as we expect

scbedd committed:
update the default build so we can see what we're credscanning

scbedd committed:
ensure that sourceversion works properly

scbedd committed:
properly fix this thing

scbedd committed:
multiline clone error

scbedd committed:
Rename CheckoutTarget to something more explicit

scbedd committed:
ensure that we properly get the full ref

scbedd committed:
ensure that we get the full ref when fetching

scbedd committed:
handle default case

scbedd commented on eng/basic_tag_matrix.json:
```diff
@@ -0,0 +1,7 @@
+{
+  "matrix": {
+    "Tag": [
+      "js/openai/openai_f96a922c3c"
```
This exists purely as an example. It always gets replaced on an invocation of the batch build.

scbedd committed:
SourceBranchName -> SourceBranch

weshaggard commented on eng/pipelines/jobs/credscan_matrix.yml:
```diff
@@ -0,0 +1,66 @@
+    timeoutInMinutes: ${{ parameters.TestTimeoutInMinutes }}
+
+    strategy:
+      matrix: $[ ${{ parameters.Matrix }} ]
```
Is it necessary to really do this in parallel? I suspect it isn't work the extra work and copying of files to do this. We probably can get away with just running it all serial.

scbedd replied:
We could, but `steps` are generated at compile time, and I much prefer this view from a diagnosability standpoint.

As we discussed in 1:1, I'll remove the matrix generation in favor of a simple one. That'll get rid of the extra supporting files, but keep the main advantage I was looking for here.

scbedd committed:
what does the matrix actually look like?

save progress

swap the matrix generation to the basic pshell script. no reason to maintain the big heavy stuff

recursive include accidentally

stages -> jobs

remove unused variable

remove runtime access

use the name of the param being passed. come on now

small cleanup
